### PR TITLE
MySQL compatibility: support 'is not distinct from' in join on section, pt1

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
* Support function `isNotDistinctFrom` in join on section for null-safe comparison, ref https://github.com/ClickHouse/ClickHouse/issues/53061

*Comment for reviewer:* PR contains bunch of added lines, but significant part of them related to test reference file, or to moving `FunctionTuple` into header file, essential changes in code is not really big.

Test reference is taken from the result of the same test queries from another dbms that supports `<=>` operator in JOIN ON.

*TODO*:
- [x] documentation
- [x] add more sophisticated tests (different algorithms, join_use_nulls, column name clashes , etc)
- [x] support in new analyzer (probably next PR) https://github.com/ClickHouse/ClickHouse/pull/54068
- [x] support syntax `a <=> b` and `a IS NOT DISTINCT FROM b` (next PR) https://github.com/ClickHouse/ClickHouse/pull/54067
- [ ] Implement function to support execution of operator as a regular expression in SELECT list

### CODE REVIEW

**Comment by robot-clickhouse-ci-2** [comment]
This is an automated comment for commit 08c4b378dbbbec555639ca562e1142af1a2f9dfb with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/53755/08c4b378dbbbec555639ca562e1142af1a2f9dfb/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>🔴 failure</td></tr>
</table>

File: src/Interpreters/TableJoin.cpp [hunk 1]
```diff
@@ -425,55 +430,180 @@ static void renameIfNeeded(String & name, const NameToNameMap & renames)
...
+        {
+            name = base_name + "_" + toString(i);
+            if (!names.contains(name))
+                return;
+        }
+    }
+}
+
+static ActionsDAGPtr createWrapWithTupleActions(
+    const ColumnsWithTypeAndName & source_coulmns,
```
Comment (novikd) [inline src/Interpreters/TableJoin.cpp:456]:
```suggestion
    const ColumnsWithTypeAndName & source_columns,
```

**Review by novikd — COMMENTED** [review]
LGTM

File: src/Interpreters/TableJoin.cpp [hunk 2]
```diff
@@ -135,7 +139,8 @@ void TableJoin::resetCollected()
 
 void TableJoin::addUsingKey(const ASTPtr & ast)
 {
-    addKey(ast->getColumnName(), renamedRightColumnName(ast->getAliasOrColumnName()), ast);
+    /// For USING key and right key AST is the same.
```
[the remark this answers is not in the document]
Comment (vdimir) [reply src/Interpreters/TableJoin.cpp]: Just because we specify key once: `SELECT ... FROM t1 JOIN t2 USING (key)` both left and right key defined by ASTIdentifier `key`. I added this example to comment either.

[the remark this answers is not in the document]
Comment (novikd) [reply src/Interpreters/TableJoin.cpp]: Yep, agree.

**Comment by UnamedRus** [comment]
Hm, so it's implemented as simple: tuple(null_column) = tuple(null_column)

**Comment by UnamedRus** [comment]
Does it affect performance? (ie more complex hashtable)
And potential optimizations? (lets imagine pushdown of WHERE when condition match JOIN key)

**Comment by vdimir** [comment]
> Does it affect performance? (ie more complex hashtable)

It can a bit, since for tuples `hashed` key type is used, while key type can be just UInt.

> And potential optimizations? (lets imagine pushdown of WHERE when condition match JOIN key)

Potentiality yes, especially in cases when we just do not handle added step during plan optimization.

With filter push-down it works fine:

<details>
<summary>EXPLAIN PLAN</summary>

```sql

EXPLAIN PLAN SELECT * FROM t1 JOIN t2 ON t1.key = t2.key WHERE t1.key = 1 AND t2.key = 1;

┌─explain───────────────────────────────────────────────────────────────────────────────────────┐
│ Expression ((Projection + (Before ORDER BY + )))                                              │
│   Join (JOIN FillRightFirst)                                                                  │
│     Filter (( + Before JOIN))                                                                 │
│       ReadFromMergeTree (default.t1)                                                          │
│     Filter (( + (Joined actions + (Rename joined columns + (Projection + Before ORDER BY))))) │
│       ReadFromMergeTree (default.t2)                                                          │
└───────────────────────────────────────────────────────────────────────────────────────────────┘

EXPLAIN PLAN SELECT * FROM t1 JOIN t2 ON t1.key <=> t2.key WHERE t1.key = 1 AND t2.key = 1;

┌─explain─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Expression ((Projection + (Before ORDER BY + )))                                                                                    │
│   Join (JOIN FillRightFirst)                                                                                                        │
│     Expression (( + (Convert JOIN columns + Before JOIN)))                                                                          │
│       Filter ((( + (Convert JOIN columns + Before JOIN)))[split])                                                                   │
│         ReadFromMergeTree (default.t1)                                                                                              │
│     Expression (( + (Convert joined columns + (Joined actions + (Rename joined columns + (Projection + Before ORDER BY))))))        │
│       Filter ((( + (Convert joined columns + (Joined actions + (Rename joined columns + (Projection + Before ORDER BY))))))[split]) │
│         ReadFromMergeTree (default.t2)                                                                                              │
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

```

</details>

But I assume it can break in case of `full_sorting_merge` and read in order. For support that we need to check that `tuple` is defined as monotonic and handle it properly.

Basically it works the same as with joining different types, e.g. `JOIN ON (t1.a :: UInt32) == (t2.a :: Int32)` in that case type conversion step is added

**Comment by UnamedRus** [comment]
I assume, it will also break with https://github.com/ClickHouse/ClickHouse/pull/53443

**Comment by vdimir** [comment]
> I assume, it will also break with https://github.com/ClickHouse/ClickHouse/pull/53443

Tests should catch it
